HeterogeneousCore/AlpakaInterface/interface/workdivision.h

```diff
@@ -302,6 +302,237 @@ namespace cms::alpakatools {
     const Vec extent_;
   };
 
+  /*********************************************
```

Comment:
@borzari @nothingface0 could you point me to the use cases for
```
element_index_range_in_block
element_index_range_in_block_truncated
element_index_range_in_grid
element_index_range_in_grid_truncated
```
and
```
for_each_element_in_block
for_each_element_in_block_strided
for_each_element_in_grid_strided
```
?

Regarding the latter, do you prefer the lambda approach (pass a function object or lambda to the `for_each_element_in_...` call) or the range loop approach (get the index via a range loop, as with `elements_with_stride`) ?

Reply:
## `element_index_range_in_block`

### `workdivision.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L366) and [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L391). It's then used by `element_index_range_in_block_truncated` and `element_index_range_in_grid`.

### `gpuClusterChargeCut.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/RecoLocalTracker/SiPixelClusterizer/plugins/alpaka/gpuClusterChargeCut.h#L63)

### `gpuPixelRecHits.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/RecoLocalTracker/SiPixelRecHits/plugins/alpaka/gpuPixelRecHits.h#L123)

### `gpuFishbone.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/RecoPixelVertexing/PixelTriplets/plugins/alpaka/gpuFishbone.h#L61)

### `gpuPixelDoubletsAlgos.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/RecoPixelVertexing/PixelTriplets/plugins/alpaka/gpuPixelDoubletsAlgos.h#L152)

## `element_index_range_in_block_truncated`

### `workdivision.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L433). See `for_each_element_in_block` for implicit uses.

## `element_index_range_in_grid`

### `workdivision.h`

Inside `for_each_element_in_grid_strided` [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L533).

### `gpuFishbone.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/RecoPixelVertexing/PixelTriplets/plugins/alpaka/gpuFishbone.h#L66)

### `gpuPixelDoubletsAlgos.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/RecoPixelVertexing/PixelTriplets/plugins/alpaka/gpuPixelDoubletsAlgos.h#L159)

## `element_index_range_in_grid_truncated`

### `workdivision.h`

Only used from overloaded function [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L424). **Not used anywhere else, can be deleted.**

---------------

## `for_each_element_in_block`

### `workdivision.h`

Call from overloaded function [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L455).

### `radixSort.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L110), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L127), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L145), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L145), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L145).


## `for_each_element_in_block_strided`

### `workdivision.h`

Call from overloaded function [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L522).

### `radixSort.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L25), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L32), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L35), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L38), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L51), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L57), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L60), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L63), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L95), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L99), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L103), and [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/radixSort.h#L211).
  
## `for_each_element_in_grid_strided`

### `workdivision.h`

Call from overloaded function [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L561). 

### `HistoContainer.h`

[here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/HistoContainer.h#L28), [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/HistoContainer.h#L47) and [here](https://github.com/borzari/cmssw/blob/PixelTrack_Alpaka/HeterogeneousCore/AlpakaUtilities/interface/HistoContainer.h#L243).

## Comments
Since we didn't have the time to go in-depth in the `patatrack-standalone` algorithms, we did not try to change different ways of looping over the elements so that they're more uniform.
That said, personally speaking, the range approach (`elements_with_stride`) seems more intuitive, and it's cleaner in the code, too. 

An overview of the dependencies of those functions can be seen [here](https://drive.google.com/file/d/1SHZAu7d2JFIH4QstXejTbMdnm6o-SD7j/view?usp=sharing) for easier(?) comprehension.